### ISSUE TITLE
Idea: Lint as you type interval in both directions

### ISSUE THREAD

**Issue by xenithorb** [steelbrain/linter#1236]
After several days of trying to get used to Atom and fighting with this linting addon, I got fed up enough to look at the settings for a delay. You see the problem that I was having is that as I'm writing code the UI freaks out and starts warning me about unfinished things as I'm stil typing.

In the settings, "Lint As You Type Interval" is what I found, and I set it to 3000ms 

That greatly improved my ability to continue to type without being distracted by the linter bar yelling at me for not finishing a piece of code yet, but on the flip side, it also didn't provide useful feedback.

Yes I do want the best of both worlds :)

What I propose is that there should be two "Lint as you type" intervals. The general idea is that the linter should allow me a reasonable amount of time to finish typing (as to not yell at me and jumble up the screen while I'm in the middle of doing something that it's yelling at me for!) and also to provide instantaneous feedback when something has been fixed (so I don't have to wait an equal amount of time to see if a quick fix made the error warning go away)

I think in this case, the linter should continue to lint at whatever interval is default, but only _*display*_ an error/warning in the event of passing said additional interval. 

Let's be conservative in the time that people are warned about issues, but liberal in getting rid of them as quickly as possible

Thank you for creating this :)

**Comment by steelbrain** [comment]
The v2 (master) fixes this. The way it fixes this is by providing users a choice to have tooltips show when you put your editor cursor or move over it. If you change it to mouse, you will see the errors but the annoying tooltip will not show.

Sounds good?

**Comment by Arcanemagus** [comment]
By "linter bar" I'm guessing they mean the panel, not the tooltips 😉.

Personally I'd say this is a _lot_ of extra complexity for something that isn't that much benefit.

**Comment by steelbrain** [comment]
God, I suck at understanding humans. Thanks @Arcanemagus. I could look into implementing this behavior change with enough votes

**Comment by xenithorb** [comment]
I disagree that it doesn't add much benefit 

The point is to make the functionality less annoying while you type. A delay interval was a good step in the right direction, but it too becomes a pain when you also have to wait the same amount of time for it to recognize your changes..

I think it would simply be better to have it lint all the time and only **display** errors (in the bottom bar pictured in screenshot) at that interval instead of it being a two-way street.
![image](https://cloud.githubusercontent.com/assets/10273878/18398602/8e805ef2-769b-11e6-9f99-66f480c49737.png)

For example, if I set this hypothetical new  interval to be 3000ms, then in the background the linter continues to lint at whatever default interval you choose, but only **displays** the errors (as pictured in the screenshot) after the last interval post-typing (typing stops). 

This way, you can fix an error and it will go away quickly, but pointlessly displayed errors from mid-typed code are not distracting the user.

**Comment by joeroback** [comment]
:+1: i agree with @xenithorb, something along these lines would improve usability, IMHO

**Comment by bradybus** [comment]
I'm fairly new to GitHub. Is there a button to vote, or does my replying do the job? I'd like to see something along these lines, too.

Thanks!
For the record, this package has helped a ton!

**Comment by Arcanemagus** [comment]
https://github.com/steelbrain/linter-ui-default/pull/427 should have (again) fixed the tooltips showing up as you are typing, is that sufficient to cover what you are requesting here, or do you literally not want new issues to show up _at all_ until this new secondary delay has passed?

If that is the case I think this issue actually belongs on `linter-ui-default`. From what I understand you are looking for `linter` to continue getting message updates at it's normal timeframe, but want the _display_ of new messages to be delayed past a certain interval. As `linter` has _nothing_ to do with the display of messages I think this should be refiled over on `linter-ui-default`, unless I'm mistaken?

**Comment by bradybus** [comment]
Yes, @Arcanemagus, I think your description is correct. I'm wondering now if a tweak to the stylesheet would be possible to change this display time. I'll have to check into this.

**Comment by skylize** [comment]
@Arcanemagus I want a longer delay before reporting errors at all. Even if an error is way down at the bottom of the page, it draws my eyes away from my work and triggers a slight twinge of "Oh no! I broke my code!" even though it is probably a meaningless error since I haven't even finished typing my thought. But once I have fixed an error, I don't want to have to wait a long time for validation. "Wait that should have worked. Why do I still have red lines?!?!........ Oh yeah ok."

By somehow splitting `Lint on Change` into `Lint Report Errors on Change` and `Lint Report OK on Change`, each running on a different interval, we could get the best aspects a short delay and the best aspects of a long delay.

I don't know how the module hierarchy works for atom linter, but I would think this issue belongs to whichever package provides the core functionality of the "`Lint on Change Interval`" as it is shown in the core `linter` Settings page.

**Comment by Arcanemagus** [comment]
`linter` is what manages running the providers and maintaining the current message set. It is what handles the `Lint on Change` interval.

`linter-ui-default` is what manages _all_ display of information, `linter` itself has _no_ user visible parts.

There are three ways this could be solved as I see it at least:

----

`linter` implements multiple delays, and only sends messages onto UI providers once they have passed the appropriate threshold delay.

Advantages:
* Generic for all UIs that hook into `linter` (`linter-ui-default`, `linter-ui-plus`, `minimap-linter`, ...?)

Disadvantages:
* No way to opt out of this for a UI provider. If you want instant results you are out of luck.
* Moves logic related _only_ to the display of messages into the management of the messages, where it doesn't belong

----

`linter-ui-default` implements a delay in displaying new messages to the user.

Advantages:
* Display logic stays in the package(s) related to displaying the messages.
* Delay configuration is obviously part of the UI
* No need to change Linter as it can continue managing providers as normal

Disadvantages:
* Each UI provider would need to implement this, if it wanted to.

----

Completely rewrite the way the Linter API works, providing different sets of messages to UI providers for display and updates of messages.

Advantages:
* It's another way of doing this?

Disadvantages:
* Linter and all UI providers would have to be rewritten
* No difference from either of the above two methods

----

If this is going to be done I would say option 2 is the best path forward as it keeps the separation of management and display of messages intact and allows the greatest flexibility in UI providers.

**Comment by skylize** [comment]
---

@Arcanemagus I think option 1 looks best. Let me explain why by marking up your analyses:

---
1. `linter` implements multiple delays, and only sends messages on to UI providers once they have passed the appropriate threshold delay.

Advantages:

* Generic for all UIs that hook into `linter` (`linter-ui-default`, `linter-ui-plus`, `minimap-linter`, ...?)

    * This is a very big advantage. Nothing more needs to be said about that. 😉

Disadvantages:

* No way to opt out of this for a UI provider. If you want instant results you are out of luck.
Moves logic related only to the display of messages into the management of the messages, where it doesn't belong

    * There is no way for UI provider to opt out of a single delay either. If user wants instant results, she can set delay to instant. If UI provider wants instant, who cares?

    * The same would hold for 2 delays: if user wants only instant results, then set both delays  to instant and all results are instant. If a UI provider wants only instant results, who cares?

    * If you really think a UI provider should be able to opt out, you can always provide an additional hook they can use to automatically sync both delays despite user settings.

* Moves logic related only to the display of messages into the management of the messages, where it doesn't belong

    * That is a valid position to consider. However, assuming it is true, I think it is a different issue than the one being discussed. At present, the delay is managed by `linter`. No way a UI can get messages before `linter` says it can have them. If you want `linter-ui-default` to take over delay management, my opinion is you should open a new issue to chalk over that decision and not handcuff this feature unnecessarily to a core structural change.
   
     * Delay management is not a purely visual exercise. If user sets a long delay then `linter` can save CPU cycles by not linting very often. If user sets short delay, then `linter` needs to keep up by linting more often. 

    * Likewise, if user has 2 intervals available, and sets one high to show errors slowly and one low to hide resolved errors quickly , then `linter` can lazily run occasional lints while no errors are displayed and then start churning out new lints as fast as possible when something looks wrong.

     * If you make UI providers implement their own delay management, then `linter` must lint continuously always and forever to make sure there is always a fresh response available when the UI asks for it. 

    * If for some odd reason `linter` is not already utilizing the cpu saving advantages of controlling the delay, then it's still available as a future optimization as long as `linter` still retains control of delay management (or offloads the job to a separate and distinct `linter-delay` module.).

---
2. `linter-ui-default` implements a delay in displaying new messages to the user.

Advantages:

* Display logic stays in the package(s) related to displaying the messages.

    * That assumes this is only about display.

* Delay configuration is obviously part of the UI

    * "Obviously" part of UI? Why is it so obvious? If it's so obvious, why is `linter` currently in charge of the delay? 

    * Please see my comments on Option 1. I think you could just as easily say delay configuration is _obviously NOT_ part of the UI. 

* No need to change `linter` as it can continue managing providers as normal

    * Well for the feature to be implemented, you either gotta write it here or or write it there. Kicking the can down the road doesn't help anybody.

Disadvantages:

Each UI provider would need to implement this, if it wanted to.

  * Yes that is a very big disadvantage. You lose the currently expected _behavioral_ consistency across providers because some developers will be too lazy (or more likely too busy) to implement a decent version of even a single delay.)

---
3. Completely rewrite the way the Linter API works, providing different sets of messages to UI providers for display and updates of messages.

Advantages:

* It's another way of doing this?
    * Ummm.... Yes?

Disadvantages:

* Linter and all UI providers would have to be rewritten

    * Ouch! 🤕

* No difference from either of the above two methods.

    * Well it is _different_, but I think it captures the worst aspects of both.

**Comment by Arcanemagus** [comment]
# Option 1

> This is a very big advantage.

Yep, and it's the only reason I'm not just flat out saying Option 2 should be the way this is done. There are just as big reasons Option 1 shouldn't be the way to go though.

> There is no way for UI provider to opt out of a single delay either. If user wants instant results, she can set delay to instant. If UI provider wants instant, who cares?

> The same would hold for 2 delays: if user wants only instant results, then set both delays to instant and all results are instant. If a UI provider wants only instant results, who cares?

I'm not sure what you are talking about here. First of all the "UI" and "user" here are the **same** as the user only ever interacts with the UI.

As a note: The `lintOnChangeInterval` should _never_ be set to 0 (instant), that would be insane. In fact there really should be a minimum specified there of something like 150 ms as even that is pretty crazy to request updates that quickly.

> If you really think a UI provider should be able to opt out, you can always provide an additional hook they can use to automatically sync both delays despite user settings.

That's my entire point here, there is no way to do this if option 1 is chosen.

> If you want linter-ui-default to take over delay management

Absolutely not, option 1 has `linter` handling both delays, option 2 has `linter` handling the delay that it currently does (`lintOnChangeInterval`), while `linter-ui-default` handles the new delay (`newMessageDelay`).

> Delay management is not a purely visual exercise. If user sets a long delay then linter can save CPU cycles by not linting very often. If user sets short delay, then linter needs to keep up by linting more often.

`linter` itself does very little actual work, the providers are where all linting (and CPU load) happens. That being said, that's exactly what `lintOnChangeInterval` is for, and has nothing to do with this discussion other than the fact that your proposed delay would make longer `lintOnChangeInterval` delays feel _even longer_ to the user.

> Likewise, if user has 2 intervals available, and sets one high to show errors slowly and one low to hide resolved errors quickly , then linter can lazily run occasional lints while no errors are displayed and then start churning out new lints as fast as possible when something looks wrong.

All three options support this, after all that's the entire point of this discussion. It almost sounds like you are suggesting a _third_ delay here though for delays between lints when there are no messages so those are farther apart than when there are messages.

> If you make UI providers implement their own delay management, then linter must lint continuously always and forever to make sure there is always a fresh response available when the UI asks for it.

UI's don't ask for responses, `linter` tells them when there are results available. The minimum delay between lints is controlled by `lintOnChangeInterval`. This has no bearing on this discussion.

> If for some odd reason linter is not already utilizing the cpu saving advantages of controlling the delay, then it's still available as a future optimization as long as linter still retains control of delay management (or offloads the job to a separate and distinct linter-delay module.).

I have no idea what you are talking about here 😕.

## Implementation

Option 1 is proposing adding a second delay option to `linter` itself:
* `lintOnChangeInterval`: The minimum amount of time between requests to a provider for messages on editor changes
* `newMessageDelay`: Delay between when a new message arrives from a provider, and when it gets sent on to the UIs

When a change is detected in an editor Linter would:
1. Check whether associated providers have been called in the last `lintOnChangeInterval` ms, if so, do nothing.
2. If they haven't, request a `lint()` from them.
3. When the updated messages from the provider arrive then:
  1. Check if any currently held messages are no longer listed, if so first check if they have ever been sent to the UI providers. If they haven't, simply remove them from the registry. Cancel any pending timed callbacks on messages. If they have previously been sent to the UI providers add them to the `removed` part of the `MessagePatch`.
  2. Check if any messages from the provider are new, if so mark them as only eligible for sending on after `newMessageDelay` ms. Add a callback that fires after `newMessageDelay` to _potentially_ send the message on to UI providers.
  3. Run through the list of current messages, checking if there are any that are past the `newMessageDelay` that have not previously been sent to the UIs, if any are found add them to the `added` part of the `MessagePatch` and mark them as sent to the UI.
  4. Filter the current messages for ones that are eligible for display, and add all of them to the `messages` list for the `MessagePatch`.
4. Send the generated `MessagePatch` to UI providers.
5. When the callback from 2. fires, send the message on to UI providers.

# Option 2

> That assumes this is only about display.

This **is** only about display, which seems to be a fundamental misunderstanding on your part.

> "Obviously" part of UI? Why is it so obvious?

By that I meant the setting would be visible under `linter-ui-default` in the settings menu, instead of `linter`. With Option 1 the setting for when a message actually showed up in a UI would be there... when _no_ other setting for the UI is there.

> If it's so obvious, why is linter currently in charge of the delay?

There is no delay currently, so I'm not sure what you mean?

With this option `linter` would remain the exact same.

## Implementation

`linter-ui-default` would need to have a new option (`newMessageDelay`) added. The code would have to be modified so its `render` function handles `MessagePatch`s differently.

* The messages listed in `added` would instead be put in a `pendingMessages` bucket, and a callback would be triggered for `newMessageDelay` ms later. They will need to be tagged with a time when they are valid.
* Messages in `removed` would be removed from the `pendingMessages` bucket as well as the main store of messages
* When the callback fires it checks if any messages are in the `pendingMessages` bucket that are now valid for display, if so it removes them from the bucket and renders them.

# Option 3

Mainly a joke, it was just a possible solution I came up with while writing the other two.

**Comment by skylize** [comment]
> I'm not sure what you are talking about here. First of all the "UI" and "user" here are the same as the user only ever interacts with the UI.

My use of the word 'user' here specifically refers to a person using Atom with linter installed. To be more clear, I will use capital U User as a proper-noun to refer to a person from here forward.

From the perspective of `linter`, a UI is the user. But from User's perspective, `linter` and `linter-ui-*` are all just a single application providing a linting service. 

When I say who cares if a UI can opt out, I mean that if somebody takes the time to build this great feature, then ideally everyone should get to use it. As long as User has a way to opt out, I don't care if UI can opt out. From what you described so far, I expect that if User opens `linter` settings page and sets `Lint on Change Interval`, that goes directly to `linter`, and not to `linter-ui-default`. Am I wrong? So the ideal solution from perspective of User is one where a second number shows up right next to the current one and it all gets handled together. Obviously though, User's ideal does not necessarily coincide with "solution that will actually lead to this being a real live feature."

> linter itself does very little actual work, the providers are where all linting (and CPU load) happens.

Ok, but what does that process look like? Are providers linting continuously and messages from them just being ignored until Lint on Change? Are providers linting only when there is empty queue for them to send messages into? Are providers waiting for `linter` to request lint?

It sounds to me like `linter` is making requests for linting every `lintOnChangeInterval` ms. So with higher value, it will do less frequent lints and thus burn less CPU. Even if the CPU usage is technically by the provider, it's the request from `linter` that triggered those cycles. 

"The lintOnChangeInterval should never be set to 0 (instant), that would be insane" because it would burn the crap out of your CPU or bring Atom to a screeching halt. Haha

With that in mind, we could toggle between using `lintOnChangeInterval` and `newMessageDelay` (ideally renaming the variables at some point to match the new duel-interval paradigm). Whenever we run out of messages to display, we can switch to `newMessageDelay` for control linting speed. Once an error shows up, switch back `lintOnChangeInterval` until we clear the queue again.

> Mainly a joke, it was just a possible solution I came up with while writing the other two.

Yes, I could tell Option 3 was _mostly_ a joke. I almost didn't include it, but instead replied with responses that were _mostly_ a joke too. Laughing is healthy.

> This is only about display, which seems to be a fundamental misunderstanding on your part.

I think once you read my implementation below, you will understand why I don't think this is about display. I am in fact suggesting a change to how often lints are requested. Any delays in display are only a byproduct of lintOnChangeInterval, just as they are now.

---
I think I understand enough now about how `lintOnChangeInterval` is used to propose a simple implementation. Here's what I envision:
```
 // shorter delay when we already have messages to target finding removals quickly
let locRemoveInterval = 200 // from User settings

 //  longer delay for when any message is a new message, no removals coming
let locNewMsgInterval: 1100 // from User settings

 //  this is now a dynamic value that will change often at runtime
let lintOnChangeInterval = lintOnChangeRemoveInterval // will not show in settings
```
1. Check whether associated providers have been called in the last `lintOnChangeInterval` ms, if so, do nothing.
2. If they haven't, request a lint() from them.
3. When the updated messages from the provider arrive then:

    i. Do everything we normally do with a new list of messages.

    ii. Check if list of remaining messages is empty. (Do we still have a full list in scope 
        here? If not, maybe we can query uis for what they've got.).

    ----- If list is empty, then set `lintOnChangeInterval = locNewMsgInterval`

    ----- If list is not empty, then set `lintOnChangeInterval = locRemoveInterval`

    iii. return to Step 1

All we're doing is checking our message count. Depending if list is zero or non-zero length, we toggle the value of `lintOnChangeInterval` between 2 numbers. This will cause differences in lint request times.

**Comment by skylize** [comment]
@Arcanemagus I just re-read the OPs vision of this feature, and if we used his vision as the template, then this would definitely be a question of how to display. I think that may be where some of the confusion came from between you and me.

My impression after reading several times is that OP was hoping for a keypress listener to suppress messages, which is a different take on the same goal. For his suggestion,  UI provider would certainly be the appropriate place to manage the event listeners and message suppressions. And linter core would need to just keep doing what it does, oblivious to the fact that half the messages sent were disappearing.

I think my idea seems a lot simpler for almost the same result, and can be built right into `linter` instead of pushing it downstream.

**Comment by skylize** [comment]
> As for your proposed implementation, ... you are requesting lints when nothing has happened to the document.

Sorry, I forgot to copy the following line from your of Opt1 proposal.

> When a change is detected in an editor Linter would:

Step iii should read return to Step 0 and wait for changes.
